fix(katex): Allow CJK characters adjacent to math delimiters

tth37 opened the pull request:
# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) and describe your changes before submitting a pull request.

**Before submitting, make sure you've checked the following:**

- [x] **Target branch:** Please verify that the pull request targets the `dev` branch.
- [x] **Description:** Provide a concise description of the changes made in this pull request.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [ ] **Documentation:** Have you updated relevant documentation [Open WebUI Docs](https://github.com/open-webui/docs), or other documentation sources?
- [ ] **Dependencies:** Are there any new dependencies? Have you updated the dependency versions in the documentation?
- [x] **Testing:** Have you written and run sufficient tests to validate the changes?
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Prefix:** To clearly categorize this pull request, prefix the pull request title using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

The current regex patterns reject math delimiters (`$...$`, etc.) when they appear directly against Chinese characters (e.g., `$A$等于$B$`). This is particularly problematic because that many LLMs naturally output math expressions adjacent to Chinese text **without spacing** (e.g., `这里，$A$是一个矩阵，$E$是单位矩阵`, more examples discussed in #13487). However, our katex validation rules require **special characters/whitespace around delimiters**, causing Chinese+math combinations failed to render.

(Notably, in native [`marked-katex-extension`](https://www.npmjs.com/package/marked-katex-extension), this case is correctly handled.)

![owui](https://github.com/user-attachments/assets/a6754ce0-e2ba-468b-bf89-cd8e5796885a)

### Changed

- Introduced `ALLOWED_SURROUNDING_CHARS` constant
- Added Unicode `\p{Script=Han}` support, updated the parsing logic in lookahead checks  (`$x$中文`) and preceding checks (`中文$x$`)
- Maintained all existing punctuation/whitespace rules

### Testings

![chinese](https://github.com/user-attachments/assets/98704a12-b4ec-49e6-a4aa-f7d2b77ad424)

![japanese](https://github.com/user-attachments/assets/d02c96f5-7d55-428e-a1b7-cf986700d3a3)

![korean](https://github.com/user-attachments/assets/4771589c-a92c-4e0f-b6f0-60fdff772e8e)

I've also conducted some regression tests, and they all passed.

---

### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

tth37 committed:
fix(katex): Allow Chinese characters adjacent to math delimiters

tth37 committed:
enh: Support more languages

tjbck commented:
is this thoroughly tested? we don't want any regression here.

tth37 commented:
Here is the report on regression testing I conducted:  

I used an LLM to generate 50 segments of Markdown+LaTeX text that do not contain CJK characters. Then, I parsed the math delimiters in these segments using both the `legacyKatexExtension` and `katexExtension` to ensure the parsing results were consistent. **I hope this regression test can confirm that, when no CJK characters are present in the text, the new `katexExtension` exhibits fully consistent parsing behavior with the `legacyKatexExtension`.**  

You can review my test code at [index.js](https://github.com/tth37/katex-regression-test/blob/master/index.js) and the 50 test cases I generated at [test-cases.json](https://github.com/tth37/katex-regression-test/blob/master/test-cases.json).

![image](https://github.com/user-attachments/assets/60d09c52-b54c-4400-8d1d-26d26e1a9906)

Let me know if you'd like any refinements!

tjbck commented:
Could you confirm with examples from https://github.com/open-webui/open-webui/discussions/3581 as well?

tth37 commented:
Yes! After reviewing the discussions, I can confirm that this PR **addresses most of the issues listed in #12881, #9273, #10281 and #3581**. However, some issues were not resolved because they align with the intended behavior (which is consistent with the original katex extension):

1. #9273: Inline formulas in formats like `xxx\(\alpha\)zzz` or `xxx$\alpha$zzz` are note rendered correctly.
  This is intentional behavior, we assume that there should be whitespace between English characters and math delimiters, and this scenario is rather rare. Rendering such cases could lead to issues, particularly when the dollar sign (`$`) is used in non-math contexts (as mentioned in #3930).

2. #3581: `[ \mathbf{\hat{a}} ]` is not rendered.
  This is also intended behavior. Single `[` and `]` are not valid math delimiters. The correct syntax for block math is `\[` and `\]`, which should work as expected.

tjbck commented:
Thank you! Appreciate your hard work as always!
